### ISSUE TITLE
Linux: "no such process" and fail to stop on breakpoints if start / stop debug session for many times

### ISSUE THREAD

**Issue by iamzhout** [go-lang-plugin-org/go-lang-idea-plugin#2218]
I have checked and this should be different problem to https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1917

My development environment:
* golang plugin version: 0.10.998
* OS: Ubuntu 14.04
* IDE: PyCharm Community Edition 5.0.3

Step to reproduce:
* $ go get github.com/astaxie/beego
* $ go get github.com/beego/bee
* $ bee api apiproject    // this will create project under GOPATH/src/apiproject
* create go project in PyCharm for above apiproject, set breakpoint in main.go and run debug for the program

The problem is that the breakpoint will be **OK** for the **FIRST** time, then if stop the debug session and start again, it will always say "no such process". 
Unless quit PyCharm and start again, again **OK** for the first time, and then **NOK** afterwards.

![20160107_dlv_debug_fail_to_set_breakpoint](https://cloud.githubusercontent.com/assets/14846203/12170482/cc74eff6-b57c-11e5-9858-e44cf508414d.png)

/Tao

**Comment by zolotov** [comment]
Look like duplicate of https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/2016 which is depends on https://github.com/derekparker/delve/issues/150

**Comment by iamzhout** [comment]
Seems not the same to me, as https://github.com/derekparker/delve/issues/150 says about `println` in golang, but here there is no such function call.
and this problem could be reproduced everytime, seems quite wired, as the previous debug session has been killed, 

and there is **NO** such problem if using dlv command line. 
* go build apiproject
* dlv exec ./apiproject
* b main.go:11
* continue
this does stop the program every time.

**Comment by dlsniper** [comment]
@iamzhout how do you stop the application when running via delve? I'm not able to fully reproduce this correctly to see if it's a problem from delve and not from the plugin (or vice-versa). Thanks

**Comment by iamzhout** [comment]
@dlsniper sorry I don't say it very clear, what I mean is that I just tried to set break point via delve command line, it do works. and if I type **exit**, and then repeat the below steps 2-6 again, the breakpoint does works fine.
1. go build apiproject
2. dlv exec ./apiproject
3. b main.go:11
4. b main.go:15
5. continue
6. continue
7. exit

I just **suppose** that if I stop the debug session in PyCharm. It also stops the whole delve command; after then I I restart the debug session, there will be another delve process. I also have checked that the process for dlv is different for different debug session.

I do noticed that the way to start dlv in PyCharm is cmdline as below, 
```
/root/.PyCharm50/config/plugins/Go/lib/dlv/linux/dlv --listen=localhost:32873 --headless=true exec /tmp/apiproject(app)0go --
```
which is not the same with ```dlv exec ./apiproject```, 

but I still suspect it is the problem resides in golang plugin, otherwise it is hard to explain why if I close and open PyCharm and then start the debug session, the breakpoint always works *ONLY* for the very first time.

**Comment by iamzhout** [comment]
After doing more test, seems the information I provided previously is misleading and **NOT** correct :(
the problem doesn't occur for the second time for debug session, but indeed more randomly.
I have tried stop and start the debug session just now for tens of times, and it sometimes will occur, sometimes will not. 

and sometimes it will print below message in PyCharm console, and the *apiproject* program still will be running in Linux OS though debug session terminated (as apiproject is somewhat like a web server and listen on port 8080)
```
process must be stopped in order to kill it
```

In summary, for now, I only can confirm that if we try to start / stop the debug session for more than 10 times, the issue will be reproduced, at least for me during my test.

**Comment by ignatov** [comment]
@iamzhout thanks for your report, I'm going to reproduce the issue.
If you can provide a small project for it will be cool.

**Comment by ignatov** [comment]
@iamzhout Can't reproduce on Mac OS X. There is no dlv process after detaching from the debug session. 
@dlsniper Please try to reproduce on Linux.

**Comment by dlsniper** [comment]
- I've created a sample project as described and I've started a debugging version of it.
- After the breakpoint was hit, I've resumed program execution and then made a request to: http://localhost:8080/v1/object which worked as expected.
- Then I've stopped the project and started the debugging session again and I've got this:

```
/usr/local/go1.5/bin/go build -o /tmp/demobego0go -gcflags "-N -l" /home/florin/golang/src/apiproject/main.go
/home/florin/IdeaProjects/google-go-lang-idea-plugin/build/idea-sandbox/plugins/Go/lib/dlv/linux/dlv --listen=localhost:35123 --headless=true exec /tmp/demobego0go --
2016/01/10 18:20:15 [parser.go:61] [I] /home/florin/golang/src/apiproject/controllers has not changed, not reloading 
2016/01/10 18:20:15 [parser.go:61] [I] /home/florin/golang/src/apiproject/controllers has not changed, not reloading 
2016/01/10 18:20:15 [asm_amd64.s:1721] [I] http server Running on :8080
2016/01/10 18:20:15 [asm_amd64.s:1721] [C] ListenAndServe: %!(EXTRA *net.OpError=listen tcp :8080: bind: address already in use)
```

Once I kill the existing process I am indeed able to replicate the issue described by @iamzhout, and the debugger indeed doesn't stop on the indicated breakpoint. However, I've noticed that the same issue doesn't happen when having two breakpoints active in the plugin: DlvVm.java:48 and DlvVm.java:60 (the Log.Info for IN and OUT). Maybe the `continue` command is sent too fast after the breakpoint and given the location of the breakpoint (right after main starts) delve doesn't have enough time to register it?

Additionally, I've noticed that the stop command doesn't actually send any detach command to delve to stop the debugged process. From what I read in the API, delve should receive a: detach command with the `kill: true` parameter in order to stop the debugged process and then exit itself. I've tried to add the command myself but I've failed. I'll open a separate issue for this.

**Comment by jimhorng** [comment]
the same happened on 
ubuntu: 15.10
IDEA: 15.02 community
go-lang-plugin: 0.10.749

with go application start as http server

**Comment by dlsniper** [comment]
I've managed to further replicate this.

It seems that at some point in the operation delve crashes which leads to the debugged process to hang in background. Log from the plugin: http://pastebin.com/Wy577dvD

I've created the issue in delve: https://github.com/derekparker/delve/issues/408. I wonder if we can do anything on our end (such as wait for all in-flight requests before sending the continue request?)

Finally, @ignatov, would it be possible to have the STDERR from delve into the console as well? I suspect that way we could see it? Should I open a separate issue for this?

Thank you.

**Comment by pylover** [comment]
I have the same issue. breakpoints are not working at all on ubuntu 14.04 amd64, and after stopping debug:

    process must be stopped in order to kill it

**Comment by dlsniper** [comment]
@pylover please open a separate issue and add a code sample which reproduces the issue. Or add it here but with more details (what's the plugin version for example).

**Comment by pylover** [comment]
I have the same issue. breakpoints are not working at all on ubuntu 14.04 amd64, and after stopping debug:

    process must be stopped in order to kill it

**Comment by dlsniper** [comment]
fixed in #2599
